File: app/src/main/java/com/main/climbingdiary/dialog/RouteDialogCreator.java
package com.main.climbingdiary.dialog;

import android.annotation.SuppressLint;
import android.content.Context;
import android.graphics.Color;
import android.support.v4.app.DialogFragment;
import android.text.Editable;
import android.text.SpannableStringBuilder;
import android.text.Spanned;
import android.text.TextWatcher;
import android.text.style.ForegroundColorSpan;
import android.util.Log;
import android.view.View;
import android.widget.ArrayAdapter;
import android.widget.AutoCompleteTextView;
import android.widget.Button;
import android.widget.EditText;
import android.widget.LinearLayout;
import android.widget.Spinner;
import android.widget.Switch;
import android.widget.TextView;

import com.main.climbingdiary.R;
import com.main.climbingdiary.common.AlertManager;
import com.main.climbingdiary.common.GradeConverter;
import com.main.climbingdiary.common.RouteConverter;
import com.main.climbingdiary.common.preferences.AppPreferenceManager;
import com.main.climbingdiary.controller.FragmentPager;
import com.main.climbingdiary.controller.SetDate;
import com.main.climbingdiary.database.entities.AreaRepository;
import com.main.climbingdiary.database.entities.Projekt;
import com.main.climbingdiary.database.entities.Route;
import com.main.climbingdiary.database.entities.SectorRepository;
import com.main.climbingdiary.models.Alert;
import com.main.climbingdiary.models.Levels;
import com.main.climbingdiary.models.Rating;
import com.main.climbingdiary.models.SportType;
import com.main.climbingdiary.models.Styles;
import com.main.climbingdiary.models.Tabs;

import java.util.regex.Pattern;

import cn.pedant.SweetAlert.SweetAlertDialog;
import lombok.Getter;

public class RouteDialogCreator {

    @Getter
    private final Spinner stil;
    @Getter
    private final Spinner level;
    @Getter
    private final Spinner rating;
    @Getter
    private final EditText date;
    @Getter
    private final EditText name;
    private final TextView nameHeader;
    @Getter
    private final AutoCompleteTextView area;
    @Getter
    private final AutoCompleteTextView sector;
    @Getter
    private final EditText comment;
    @Getter
    private final Context _context;
    private final View view;
    private final Switch gradeSwitcher;
    private final DialogFragment fragment;
    private final LinearLayout routeContent;
    private final LinearLayout gradeSwitcherLayout;
    @Getter
    Button closeDialog;
    @Getter
    Button saveRoute;

    public RouteDialogCreator(View _view, Context context, DialogFragment _fragment) {
        view = _view;
        closeDialog = view.findViewById(R.id.input_route_close);
        saveRoute = view.findViewById(R.id.input_route_save);
        stil = view.findViewById(R.id.input_route_stil);
        level = view.findViewById(R.id.input_route_level);
        rating = view.findViewById(R.id.input_route_rating);
        date = view.findViewById(R.id.input_route_date);
        name = view.findViewById(R.id.input_route_name);
        nameHeader = view.findViewById(R.id.input_route_name_header);
        area = view.findViewById(R.id.input_route_area);
        sector = view.findViewById(R.id.input_route_sektor);
        comment = view.findViewById(R.id.input_route_comment);
        gradeSwitcher = view.findViewById(R.id.grade_system_switcher);
        routeContent = view.findViewById(R.id.route_content);
        gradeSwitcherLayout = view.findViewById(R.id.grade_switcher_layout);
        fragment = _fragment;
        _context = context;
        setOnCloseButton();
    }

    @SuppressLint("SetTextI18n")
    public void setUiElements(Projekt projekt) {

        saveRoute.setText("Update");
        name.setText(projekt.getName());
        this.setRouteNameHeaderText();
        // set Spinner for choosing the level
        this.setLevelSpinner(projekt.getLevel(), Levels.INSTANCE.getLevelsFrench());
        this.setGradeSwitcher(true);

        //get the route List and set autocomplete
        this.setAreaSectorAutoComplete();
        area.setText(projekt.getArea());
        sector.setText(projekt.getSector());
        comment.setText(projekt.getComment());
        // set the Spinner
        this.setRatingSpinner(projekt.getRating() - 1);
        new SetDate(date, _context);
        routeContent.setVisibility(View.GONE);
    }

    @SuppressLint("SetTextI18n")
    public void setUiElements(Route route) {

        saveRoute.setText("Update");
        name.setText(route.getName());
        this.setRouteNameHeaderText();
        // set Spinner for choosing the style
        this.setStilSpinner(route.getStyle().toUpperCase());
        // set Spinner for choosing the level
        this.setLevelSpinner(route.getLevel(), Levels.INSTANCE.getLevelsFrench());

        //get the route List and set autocomplete
        this.setAreaSectorAutoComplete();
        area.setText(route.getArea());
        sector.setText(route.getSector());
        comment.setText(route.getComment());
        // set the Spinner
        this.setRatingSpinner(route.getRating() - 1);
        new SetDate(date, _context);

        this.setGradeSwitcher(false);

        //set date listener
        try {
            new SetDate(date, _context);
            StringBuilder sBuilder = new StringBuilder();
            String[] dateSplit = (route.getDate()).split(Pattern.quote("."));
            sBuilder.append(dateSplit[2]).append("-").append(dateSplit[1]).append("-").append(dateSplit[0]);
            date.setText(sBuilder.toString());
        } catch (Exception e) {
            Log.e("Error", e.toString());
            date.setText(route.getDate());
        }
    }

    public void setUiElements(boolean projekt) {
        if (projekt) {
            routeContent.setVisibility(View.GONE);
        }
        this.setStilSpinner("RP");
        this.setLevelSpinner("8a", Levels.INSTANCE.getLevelsFrench());
        this.setAreaSectorAutoComplete();
        this.setRatingSpinner(1);
        this.setGradeSwitcher(true);
        this.setRouteNameHeaderText();
        new SetDate(date, _context);
    }

    public void setForeGroundSpan(String title) {
        // Initialize a new foreground color span instance
        ForegroundColorSpan foregroundColorSpan = new ForegroundColorSpan(Color.BLACK);

        // Initialize a new spannable string builder instance
        SpannableStringBuilder ssBuilder = new SpannableStringBuilder(title);

        // Apply the text color span
        ssBuilder.setSpan(foregroundColorSpan, 0, title.length(), Spanned.SPAN_EXCLUSIVE_EXCLUSIVE);

        fragment.getDialog().setTitle(ssBuilder);

        // Set title divider color
        int titleDividerId = fragment.getResources().getIdentifier("titleDivider", "id", "android");
        View titleDivider = fragment.getDialog().findViewById(titleDividerId);
        if (titleDivider != null)
            titleDivider.setBackgroundColor(fragment.getResources().getColor(android.R.color.black));
    }

    private void setGradeSwitcher(boolean visibility) {
        int viewState = visibility && AppPreferenceManager.getSportType() == SportType.KLETTERN ? View.VISIBLE : View.GONE;

        gradeSwitcherLayout.setVisibility(viewState);
        gradeSwitcher.setOnCheckedChangeListener((buttonView, isChecked) -> {
            if (isChecked) {
                this.setLevelSpinner("8a", Levels.INSTANCE.getLevelsFrench());
            } else {
                this.setLevelSpinner("IX+/X-", Levels.INSTANCE.getLevelsUiaa());
            }
        });
    }

    private void setAreaSectorAutoComplete() {
        ArrayAdapter<String> areaArrayAdapter = new ArrayAdapter<String>(view.getContext(), android.R.layout.simple_spinner_item, AreaRepository.INSTANCE.getAreaNameList());
        //will start working from first character
        area.setThreshold(1);
        area.setAdapter(areaArrayAdapter);
        area.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {

            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                ArrayAdapter<String> sectorArrayAdapter = new ArrayAdapter<String>(_context, android.R.layout.simple_spinner_item, SectorRepository.INSTANCE.getSectorList(area.getText().toString().trim()));
                //will start working from first character
                sector.setThreshold(1);
                sector.setAdapter(sectorArrayAdapter);
            }

            @Override
            public void afterTextChanged(Editable s) {

            }
        });
    }

    private void setStilSpinner(String selection) {
        // set Spinner for choosing the style
        ArrayAdapter<String> stilArrayAdapter = new ArrayAdapter<String>(_context, android.R.layout.simple_spinner_item, Styles.INSTANCE.getStyle(true));
        stilArrayAdapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item); // The drop down view
        stil.setAdapter(stilArrayAdapter);

        try {
            stil.setSelection(stilArrayAdapter.getPosition(selection));
        } catch (Exception ex) {
            Log.e("Style was unset", ex.getLocalizedMessage());
        }
    }

    private void setLevelSpinner(String selection, String levels[]) {
        ArrayAdapter<String> levelArrayAdapter = new ArrayAdapter<String>(_context, android.R.layout.simple_spinner_item, levels);
        levelArrayAdapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item); // The drop down view
        level.setAdapter(levelArrayAdapter);
        level.setSelection(levelArrayAdapter.getPosition(selection));
    }

    private void setRatingSpinner(int selection) {
        // set Spinner for choosing the level
        ArrayAdapter<String> ratingArrayAdapter = new ArrayAdapter<String>(_context, android.R.layout.simple_spinner_item, Rating.INSTANCE.getRating());
        ratingArrayAdapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item); // The drop down view
        rating.setAdapter(ratingArrayAdapter);
        rating.setSelection(selection);
    }

    private void setOnCloseButton() {
        //close the dialog
        getCloseDialog().setOnClickListener(v -> {
            if (AppPreferenceManager.getSelectedTabsTitle() == Tabs.PROJEKTE) {
                FragmentPager.getInstance().refreshSelectedFragment();
            }
            fragment.getDialog().cancel();
        });
    }

    private void setRouteNameHeaderText() {
        String header = AppPreferenceManager.getSportType().getRouteName() + "name";
        nameHeader.setText(header);
    }

    public Route getRoute(boolean id) {
        Route newRoute = new Route();
        if (id) {
            newRoute.setId(0);
        }
        String level = this.getLevel().getSelectedItem().toString();
        try {
            level = gradeSwitcher.isChecked() ? level : GradeConverter.convertUiaaToFrench(level);
        } catch (Exception ignored) {
        }
        newRoute.setName(this.getName().getText().toString());
        newRoute.setLevel(level);
        newRoute.setDate(this.getDate().getText().toString());
        newRoute.setArea(this.getArea().getText().toString());
        newRoute.setSector(this.getSector().getText().toString());
        newRoute.setComment(this.getComment().getText().toString());
        newRoute.setRating(this.getRating().getSelectedItemPosition() + 1);
        newRoute.setStyle(this.getStil().getSelectedItem().toString());
        return (Route) RouteConverter.cleanRoute(newRoute);
    }

    public Projekt getProjekt(boolean id) {
        Projekt projekt = new Projekt();
        if (id) {
            projekt.setId(0);
        }
        String level = this.getLevel().getSelectedItem().toString();
        try {
            level = gradeSwitcher.isChecked() ? level : GradeConverter.convertUiaaToFrench(level);
        } catch (Exception ignored) {
        }
        projekt.setName(this.getName().getText().toString());
        projekt.setLevel(level);
        projekt.setArea(this.getArea().getText().toString());
        projekt.setSector(this.getSector().getText().toString());
        projekt.setComment(this.getComment().getText().toString());
        projekt.setRating(this.getRating().getSelectedItemPosition() + 1);
        return (Projekt) RouteConverter.cleanRoute(projekt);
    }

    public boolean checkDate() {
        if (date.getText().toString().trim().length() == 0) {
            Alert alert = new Alert.Builder()
                    .title(String.format("Datum fehlt %s", "\ud83d\ude13"))
                    .dialogType(SweetAlertDialog.ERROR_TYPE)
                    .build();
            new AlertManager().setAlertWithoutContent(_context, alert);
            return false;
        } else {
            return true;
        }
    }
}

File: app/src/main/java/com/main/climbingdiary/dialog/RouteDialogCreator.kt
package com.main.climbingdiary.dialog

import android.annotation.SuppressLint
import android.content.Context
import android.graphics.Color
import android.text.Editable
import android.text.SpannableStringBuilder
import android.text.Spanned
import android.text.TextWatcher
import android.text.style.ForegroundColorSpan
import android.util.Log
import android.view.View
import android.widget.*
import androidx.fragment.app.DialogFragment
import cn.pedant.SweetAlert.SweetAlertDialog
import com.main.climbingdiary.R
import com.main.climbingdiary.common.AlertManager.setAlertWithoutContent
import com.main.climbingdiary.common.GradeConverter.convertUiaaToFrench
import com.main.climbingdiary.common.RouteConverter.cleanRoute
import com.main.climbingdiary.common.preferences.AppPreferenceManager.getSelectedTabsTitle
import com.main.climbingdiary.common.preferences.AppPreferenceManager.getSportType
import com.main.climbingdiary.controller.FragmentPager
import com.main.climbingdiary.controller.FragmentPager.refreshSelectedFragment
import com.main.climbingdiary.controller.SetDate
import com.main.climbingdiary.database.entities.AreaRepository.getAreaNameList
import com.main.climbingdiary.database.entities.Projekt
import com.main.climbingdiary.database.entities.Route
import com.main.climbingdiary.database.entities.SectorRepository.getSectorList
import com.main.climbingdiary.models.Alert
import com.main.climbingdiary.models.Levels.getLevelsFrench
import com.main.climbingdiary.models.Levels.getLevelsUiaa
import com.main.climbingdiary.models.Rating.getRating
import com.main.climbingdiary.models.SportType
import com.main.climbingdiary.models.Styles.getStyle
import com.main.climbingdiary.models.Tabs
import java.util.*
import java.util.regex.Pattern

class RouteDialogCreator(
    val view: View,
    val context: Context,
    private val fragment: DialogFragment
) {

    private var stil: Spinner = view.findViewById(R.id.input_route_stil)
    private var level: Spinner = view.findViewById(R.id.input_route_level)
    private var rating: Spinner = view.findViewById(R.id.input_route_rating)
    private var date: EditText = view.findViewById(R.id.input_route_date)
    private var name: EditText = view.findViewById(R.id.input_route_name)
    private var nameHeader: TextView = view.findViewById(R.id.input_route_name_header)
    private var area: AutoCompleteTextView = view.findViewById(R.id.input_route_area)
    private var sector: AutoCompleteTextView = view.findViewById(R.id.input_route_sektor)
    private var comment: EditText =
        view.findViewById<AutoCompleteTextView>(R.id.input_route_comment)
    @SuppressLint("UseSwitchCompatOrMaterialCode")
    private var gradeSwitcher: Switch = view.findViewById(R.id.grade_system_switcher)
    private var routeContent: LinearLayout = view.findViewById(R.id.route_content)
    private var gradeSwitcherLayout: LinearLayout = view.findViewById(R.id.grade_switcher_layout)
    private val closeDialog: Button = view.findViewById(R.id.input_route_close)
    var saveRoute: Button = view.findViewById(R.id.input_route_save)

    init {
        setOnCloseButton()
    }

    @SuppressLint("SetTextI18n")
    fun setUiElements(projekt: Projekt) {
        saveRoute.text = "Update"
        name.setText(projekt.name)
        setRouteNameHeaderText()
        // set Spinner for choosing the level
        setLevelSpinner(projekt.level, getLevelsFrench())
        setGradeSwitcher(true)

        //get the route List and set autocomplete
        setAreaSectorAutoComplete()
        area.setText(projekt.area)
        sector.setText(projekt.sector)
        comment.setText(projekt.comment)
        // set the Spinner
        setRatingSpinner(projekt.rating!! - 1)
        SetDate(date, context)
        routeContent.visibility = View.GONE
    }

    @SuppressLint("SetTextI18n")
    fun setUiElements(route: Route?) {
        saveRoute.text = "Update"
        name.setText(route?.name)
        setRouteNameHeaderText()
        // set Spinner for choosing the style
        setStilSpinner(route!!.style.toUpperCase(Locale.ROOT))
        // set Spinner for choosing the level
        setLevelSpinner(route.level, getLevelsFrench())

        //get the route List and set autocomplete
        setAreaSectorAutoComplete()
        area.setText(route.area)
        sector.setText(route.sector)
        comment.setText(route.comment)
        // set the Spinner
        setRatingSpinner(route.rating!! - 1)
        SetDate(date, context)
        setGradeSwitcher(false)

        //set date listener
        try {
            SetDate(date, context)
            val sBuilder = StringBuilder()
            val dateSplit = route.date!!.split(Pattern.quote(".").toRegex()).toTypedArray()
            sBuilder.append(dateSplit[2]).append("-").append(dateSplit[1]).append("-")
                .append(dateSplit[0])
            date.setText(sBuilder.toString())
        } catch (e: Exception) {
            Log.e("Error", e.toString())
            date.setText(route.date)
        }
    }

    fun setUiElements(projekt: Boolean) {
        if (projekt) {
            routeContent.visibility = View.GONE
        }
        setStilSpinner("RP")
        setLevelSpinner("8a", getLevelsFrench())
        setAreaSectorAutoComplete()
        setRatingSpinner(1)
        setGradeSwitcher(true)
        setRouteNameHeaderText()
        SetDate(date, context)
    }

    fun setForeGroundSpan(title: String) {
        // Initialize a new foreground color span instance
        val foregroundColorSpan = ForegroundColorSpan(Color.BLACK)

        // Initialize a new spannable string builder instance
        val ssBuilder = SpannableStringBuilder(title)

        // Apply the text color span
        ssBuilder.setSpan(foregroundColorSpan, 0, title.length, Spanned.SPAN_EXCLUSIVE_EXCLUSIVE)
        fragment.dialog!!.setTitle(ssBuilder)

        // Set title divider color
        val titleDividerId: Int =
            fragment.resources.getIdentifier("titleDivider", "id", "android")
        fragment.dialog!!.findViewById<View?>(titleDividerId)?.setBackgroundColor(
            fragment.resources.getColor(android.R.color.black)
        )
    }

    private fun setGradeSwitcher(visibility: Boolean) {
        val viewState =
            if (visibility && getSportType() === SportType.KLETTERN) View.VISIBLE else View.GONE
        gradeSwitcherLayout.visibility = viewState
        gradeSwitcher.setOnCheckedChangeListener { _: CompoundButton?, isChecked: Boolean ->
            if (isChecked) {
                setLevelSpinner("8a", getLevelsFrench())
            } else {
                setLevelSpinner("IX+/X-", getLevelsUiaa())
            }
        }
    }

    private fun setAreaSectorAutoComplete() {
        val areaArrayAdapter =
            ArrayAdapter(view.context, android.R.layout.simple_spinner_item, getAreaNameList())
        //will start working from first character
        area.threshold = 1
        area.setAdapter(areaArrayAdapter)
        area.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {}
            override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
                val sectorArrayAdapter = ArrayAdapter(
                    context,
                    android.R.layout.simple_spinner_item,
                    getSectorList(area.text.toString().trim { it <= ' ' })
                )
                //will start working from first character
                sector.threshold = 1
                sector.setAdapter(sectorArrayAdapter)
            }

            override fun afterTextChanged(s: Editable) {}
        })
    }

    private fun setStilSpinner(selection: String) {
        // set Spinner for choosing the style
        val stilArrayAdapter =
            ArrayAdapter(context, android.R.layout.simple_spinner_item, getStyle(true))
        stilArrayAdapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item) // The drop down view
        stil.adapter = stilArrayAdapter
        try {
            stil.setSelection(stilArrayAdapter.getPosition(selection))
        } catch (ex: Exception) {
            Log.e("Style was unset", ex.localizedMessage)
        }
    }

    private fun setLevelSpinner(selection: String, levels: Array<String>) {
        val levelArrayAdapter = ArrayAdapter(context, android.R.layout.simple_spinner_item, levels)
        levelArrayAdapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item) // The drop down view
        level.adapter = levelArrayAdapter
        level.setSelection(levelArrayAdapter.getPosition(selection))
    }

    private fun setRatingSpinner(selection: Int) {
        // set Spinner for choosing the level
        val ratingArrayAdapter =
            ArrayAdapter(context, android.R.layout.simple_spinner_item, getRating())
        ratingArrayAdapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item) // The drop down view
        rating.adapter = ratingArrayAdapter
        rating.setSelection(selection)
    }

    private fun setOnCloseButton() {
        //close the dialog
        closeDialog.setOnClickListener {
            if (getSelectedTabsTitle() === Tabs.PROJEKTE) {
                refreshSelectedFragment()
            }
            fragment.dialog!!.cancel()
        }
    }

    private fun setRouteNameHeaderText() {
        val header = getSportType().getRouteName() + "name"
        nameHeader.text = header
    }

    fun getRoute(id: Boolean): Route {
        val newRoute = Route()
        if (id) {
            newRoute.id = 0
        }
        var level: String = this.level.selectedItem.toString()
        try {
            level =
                if (gradeSwitcher.isChecked) level else convertUiaaToFrench(level).toString()
        } catch (ignored: Exception) {
        }
        newRoute.name = this.name.text.toString()
        newRoute.level = level
        newRoute.date = this.date.text.toString()
        newRoute.area = this.area.text.toString()
        newRoute.sector = this.sector.text.toString()
        newRoute.comment = this.comment.text.toString()
        newRoute.rating = this.rating.selectedItemPosition + 1
        newRoute.style = this.stil.selectedItem.toString()
        return cleanRoute(newRoute) as Route
    }

    fun getProjekt(id: Boolean): Projekt {
        val projekt = Projekt()
        if (id) {
            projekt.id = 0
        }
        var level: String = this.level.selectedItem.toString()
        try {
            level =
                if (gradeSwitcher.isChecked) level else convertUiaaToFrench(level).toString()
        } catch (ignored: Exception) {
        }
        projekt.name = this.name.text.toString()
        projekt.level = level
        projekt.area = this.area.text.toString()
        projekt.sector = this.sector.text.toString()
        projekt.comment = this.comment.text.toString()
        projekt.rating = this.rating.selectedItemPosition + 1
        return cleanRoute(projekt) as Projekt
    }

    fun checkDate(): Boolean {
        return if (date.text.toString().trim { it <= ' ' }.isEmpty()) {
            val alert = Alert("Datum fehlt \ud83d\ude13", "", SweetAlertDialog.ERROR_TYPE)
            setAlertWithoutContent(context, alert)
            false
        } else {
            true
        }
    }

}